lambda/aws-qa-appsync-opensearch/question_answering/src/qa_agent/image_qa.py

```diff
@@ -0,0 +1,258 @@
+    max_docs = input_params['retrieval']['max_docs']
+    output_file_name = input_params['filename']
+
+    source_documents = doc_index.similarity_search(decoded_question, k=max_docs)
+    logger.info(f'source is :: {source_documents}')
+    # --------------------------------------------------------------------------
+    # If an output file is specified, filter the response to only include chunks  
+    # related to that file. The source metadata is added when embeddings are 
+    # created in the ingestion pipeline.
+    # 
+    # --------------------------------------------------------------------------
+    logger.info(f'output_file_name is :: {output_file_name}')
+    if output_file_name:
+        source_documents = [doc for doc in source_documents if doc.metadata['source'] == output_file_name]
+    status_variables['sources'] = list(set(doc.metadata['source'] for doc in source_documents))
+    
+    # get new presigned url for RAG from source metadata
+    if source_documents is not None:
+        filename = source_documents[0].metadata['filename']
+        image_url = get_presigned_url(bucket_name,filename)
+        logger.info(f'image_url: {image_url}')
+        status_variables['filename']=filename
+    
+    if image_url is None:
+        logger.error(f'cannot find image url, returning..') 
+        status_variables['jobstatus'] = JobStatus.ERROR_LOAD_DOC.status
+        error = JobStatus.ERROR_LOAD_DOC.get_message()
+        status_variables['answer'] = error.decode("utf-8")
+        send_job_status(status_variables)
+        return
+    
+    # 2 : load llm using the selector
+    streaming = input_params.get("streaming", False)
+    callback_manager = [StreamingCallbackHandler(status_variables)] if streaming else None
+    #_qa_llm = get_llm(callback_manager)
+    if model_provider=='Sagemaker Endpoint':
+        _qa_llm = MultiModal.sagemakerendpoint_llm(qa_model_id)
+    elif model_provider=='Bedrock':
+        _qa_llm = get_llm(callback_manager)
+    else:
+        logger.info('Invalid Model provider, cannot load  LLM , returning..')
+        status_variables['jobstatus'] = JobStatus.ERROR_LOAD_LLM.status
+        error = JobStatus.ERROR_LOAD_LLM.get_message()
+        status_variables['answer'] = error.decode("utf-8")
+        send_job_status(status_variables)
+        return status_variables
+        
+
+    logger.info(f"going to answer the question=\"{decoded_question}\" using the image \"{image_url}\"")
+
+    # 3. Run it
+    template = """\n\nUser: {question}![]({image})<end_of_utterance>
+         \n\nAssistant:"""
+    prompt = PromptTemplate(template=template, input_variables=["image", "question"])
+    chain = LLMChain(llm=_qa_llm, prompt=prompt, verbose=input_params['verbose'])
+
+
+    try:
+        
+        logger.info(f'decoded_question is: {decoded_question}')
+        
+        tmp = chain.predict(image=image_url, question=decoded_question)
+        answer=tmp.split("Assistant:",1)[1]
```

Comment:
same question as above

Reply:
yes this is only for Idefics on sagemaker endpoint. Once tested with more model will make it more generic.